### ISSUE TITLE
DriverCreationException - invalid argument: can't kill an exited process

### ISSUE THREAD

**Issue by rameshnatesan** [mozilla/geckodriver#1500]
## System

* Version: 0.24.0
* Platform: 'Linux', os.arch: 'amd64', os.version: '2.6.32-754.9.1.el6.x86_64'
* Firefox: 60.4.0
* Selenium: 3.141.59

## Testcase

The tests are running in parallel in 8 firefox processes. I randomly see this DriverCreationException "org.openqa.selenium.WebDriverException: connection refused" when running tests in Linux machine with Mozilla Firefox 60.4.0 installed.

I even tried setting xvfb display to an offset of 5. Also, checked any defunct/zombie geckodriver and firefox process. There were no such defunct processes.

## Stacktrace 

`geb.driver.DriverCreationException: failed to create driver from callback 'script15499126431121922215376$_run_closure1$_closure4$_closure7@6c150c65'
	at geb.driver.CallbackDriverFactory.getDriver(CallbackDriverFactory.groovy:35)
	at geb.driver.CachingDriverFactory.getDriver_closure3(CachingDriverFactory.groovy:85)
	at geb.driver.CachingDriverFactory.getDriver_closure3(CachingDriverFactory.groovy)
	at geb.driver.CachingDriverFactory$SimpleCache.get(CachingDriverFactory.groovy:32)
	at geb.driver.CachingDriverFactory.getDriver(CachingDriverFactory.groovy:84)
	at geb.Configuration.createDriver(Configuration.groovy:382)
	at geb.Configuration.getDriver(Configuration.groovy:371)
	at geb.Browser.getDriver(Browser.groovy:111)
	at geb.Browser.retrieveCurrentUri(Browser.groovy:511)
	at geb.Browser.go(Browser.groovy:498)
	at geb.spock.GebSpec.methodMissing(GebSpec.groovy:56)
	at common.LoginHelper.login(LoginHelper.groovy:17)
	at tests.MySpec. user logs in successfully (MySpec.groovy:57)
### Caused by: org.openqa.selenium.WebDriverException: invalid argument: can't kill an exited process
Build info: version: '3.12.0', revision: '7c6e0b3', time: '2018-05-08T14:04:26.12Z'
System info: host: 'va1c143', ip: '10.20.10.77', os.name: 'Linux', os.arch: 'amd64', os.version: '2.6.32-754.9.1.el6.x86_64', java.version: '1.8.0_191'
Driver info: driver.version: FirefoxDriver
remote stacktrace: 
	at org.openqa.selenium.remote.W3CHandshakeResponse.lambda$new$0(W3CHandshakeResponse.java:57)
	at org.openqa.selenium.remote.W3CHandshakeResponse.lambda$getResponseFunction$2(W3CHandshakeResponse.java:104)
	at org.openqa.selenium.remote.ProtocolHandshake.lambda$createSession$0(ProtocolHandshake.java:123)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:126)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:73)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:136)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:83)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:543)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:207)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:130)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:125)
	at script15499126431121922215376$_run_closure1.closure4$_closure7(script15499126431121922215376.groovy:17)
	at script15499126431121922215376$_run_closure1.closure4$_closure7(script15499126431121922215376.groovy)
	at geb.driver.CallbackDriverFactory.getDriver(CallbackDriverFactory.groovy:29)
	... 12 more`

**Comment by btkukinom** [comment]
I have similar problems as well. `can't kill an exited process` occurs randomly.

geckodriver:0.23.0
selenium:3.14.0
firefox:60.3.0 ESR
Platform:'Linux', os.arch: 'amd64', os.version: '3.10.0-693.21.1.el7.x86_64', java.version: '11.0.2'

**Comment by whimboo** [comment]
@rameshnatesan, could you please attach a [trace-level log] from geckodriver? Read more about reporting actionable bugs in our [contribution guidelines].

@btkukinom even you see the same failure it doesn't mean that you actually have the same problem. Please file a new issue and provide all the necessary details as requested. Thanks.

[trace-level log]: https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/TraceLogs.html
[contribution guidelines]: https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/Bugs.html

**Comment by whimboo** [comment]
This issue is not about killing the process but about driver creation. As such lets use a proper description for the issue.

**Comment by ronnielivingsince1994** [comment]
Any update on this thread? I've been struggling with something similar. Not sure if creating another personal issue is the right way to go about it. 

https://github.com/buitron/geckodriver-buildpack/issues/1

https://github.com/ThoughtfulDev/EagleEye/issues/66

**Comment by whimboo** [comment]
@ronnielivingsince1994, yes please create your own issue. Thanks.

Given that we haven't gotten any further response here, I will close this issue.

**Comment by shweta8996** [comment]
what is the fix for this issue? i have same problem. i have tried all different combination of the geckodriver, firefox selenium

geckodriver 0.24.0
firefox 62.0.02
selenium 3.14

i still see the error cant kill an exited process. same thing works well on local machine its failing only when i push the docker in gitlab and run it there


from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By
from selenium import webdriver
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
from selenium.webdriver.firefox.options import Options


def before_all(context):
    # Taking input of browser name from behave.ini file and selecting the browser .
    browser_name = str.lower(context.config.userdata['browser'])

    # Selecting Firefox browser on input firefox/FIREFOX/Firefox/F/f
    if browser_name == 'firefox' or browser_name == 'f':
        try:
            #options = webdriver.FirefoxOptions()
            #options.add_argument('-headless')
            geckodriver = '/usr/local/bin/geckodriver'

            options = Options()
            options.add_argument('-headless')

            # options = Options()
            # options.headless = True
            caps = DesiredCapabilities().FIREFOX.copy()
            caps["pageLoadStrategy"] = "none"
            context._the_browser = webdriver.Firefox(executable_path=geckodriver, capabilities=caps, options=options)
            context._the_browser.implicitly_wait(10)
        except Exception as e:
            print('Error while initializing the ', browser_name, ' :--', e)

**Comment by ronnielivingsince1994** [comment]
For me it was, Geckodriver and Firefox version were not compatible. Hope it helps you.

**Comment by shweta8996** [comment]
Which version of firefox and geckodriver you used ?

**Comment by ronnielivingsince1994** [comment]
If it helps your case, I wrote my own buildpack for heroku. I didn't go for the latest versions, but rather used stable versions for both of them 
VERSION_FIREFOX=57.0.1
VERSION_GECKODRIVER=0.19.1

**Comment by whimboo** [comment]
You can always find an updated support matrix at https://firefox-source-docs.mozilla.org/testing/geckodriver/Support.html. And without a tracelog there is nothing we can do.

Also note that this issue is closed. It would be better if you could create your own one if it still persists with the latest releases, and you have all the information as requested.
